In [91]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, roc_auc_score, roc_curve
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler, PowerTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score


df = pd.read_csv(r"C:\Users\kavin\OneDrive\Desktop\End to end\cleaned_water_potability.csv")

X = df.drop(columns="Potability")
y = df["Potability"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)


scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


model = LogisticRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

print("Logistic Regression Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision: {precision_score(y_test, y_pred):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba):.4f}")


Logistic Regression Performance:
Accuracy:  0.6098
Precision: 0.0000
Recall:    0.0000
F1 Score:  0.0000
ROC-AUC:   0.5484


c:\Users\kavin\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [77]:
logreg_l1 = LogisticRegression(penalty="l1", solver="liblinear", C=1.0, random_state=42)
logreg_l1.fit(X_train, y_train)

y_pred_l1 = logreg_l1.predict(X_test)
y_proba_l1 = logreg_l1.predict_proba(X_test)[:, 1]

print("Logistic Regression (L1) Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_l1):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_l1):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_l1):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_l1):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_l1):.4f}")


Logistic Regression (L1) Performance:
Accuracy:  0.6098
Precision: 0.0000
Recall:    0.0000
F1 Score:  0.0000
ROC-AUC:   0.5485


c:\Users\kavin\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [78]:
logreg_l2 = LogisticRegression(penalty="l2", solver="lbfgs", C=1.0, max_iter=1000, random_state=42)
logreg_l2.fit(X_train, y_train)

y_pred_l2 = logreg_l2.predict(X_test)
y_proba_l2 = logreg_l2.predict_proba(X_test)[:, 1]

print("\nLogistic Regression (L2) Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_l2):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_l2):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_l2):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_l2):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_l2):.4f}")



Logistic Regression (L2) Performance:
Accuracy:  0.6098
Precision: 0.0000
Recall:    0.0000
F1 Score:  0.0000
ROC-AUC:   0.5484


c:\Users\kavin\anaconda4\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [79]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)
y_proba_knn = knn.predict_proba(X_test)[:, 1]

print("KNN Classifier Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_knn):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_knn):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_knn):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_knn):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_knn):.4f}")


KNN Classifier Performance:
Accuracy:  0.6113
Precision: 0.5031
Recall:    0.3203
F1 Score:  0.3914
ROC-AUC:   0.5991


In [89]:
svm = SVC(kernel="rbf",probability=True, random_state=42)
svm.fit(X_train, y_train)


y_pred_svm = svm.predict(X_test)
y_proba_svm = svm.predict_proba(X_test)[:, 1]


print("SVM Classifier Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_svm):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_svm):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_svm):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_svm):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_svm):.4f}")

SVM Classifier Performance:
Accuracy:  0.6692
Precision: 0.6970
Recall:    0.2695
F1 Score:  0.3887
ROC-AUC:   0.6481


In [81]:
dt = DecisionTreeClassifier(
    criterion="gini",max_depth=3,random_state=42)
dt.fit(X_train, y_train)


y_pred_dt = dt.predict(X_test)
y_proba_dt = dt.predict_proba(X_test)[:, 1]

print("Decision Tree Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_dt):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_dt):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_dt):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_dt):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_dt):.4f}")

Decision Tree Performance:
Accuracy:  0.6326
Precision: 0.6119
Recall:    0.1602
F1 Score:  0.2539
ROC-AUC:   0.5656


In [ ]:
rf = RandomForestClassifier(n_estimators=100,criterion="gini",  max_depth=None,random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
y_proba_rf = rf.predict_proba(X_test)[:, 1]


print("Random Forest Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_rf):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_rf):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_rf):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_rf):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_rf):.4f}")

Random Forest Performance:
Accuracy:  0.6555
Precision: 0.6154
Recall:    0.3125
F1 Score:  0.4145
ROC-AUC:   0.6430


In [ ]:
nb = GaussianNB()
nb.fit(X_train, y_train)


y_pred_nb = nb.predict(X_test)
y_proba_nb = nb.predict_proba(X_test)[:, 1]


print("Naive Bayes Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_nb):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_nb):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_nb):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_nb):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_nb):.4f}")

Naive Bayes Performance:
Accuracy:  0.6143
Precision: 0.5161
Recall:    0.1875
F1 Score:  0.2751
ROC-AUC:   0.6088


In [ ]:
gb = GradientBoostingClassifier(random_state=42)
gb.fit(X_train, y_train)


y_pred_gb = gb.predict(X_test)
y_proba_gb = gb.predict_proba(X_test)[:, 1]


print("Gradient Boosting Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_gb):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_gb):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_gb):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_gb):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_gb):.4f}")

Gradient Boosting Performance:
Accuracy:  0.6509
Precision: 0.6364
Recall:    0.2461
F1 Score:  0.3549
ROC-AUC:   0.6560


In [85]:
ab = AdaBoostClassifier(n_estimators=100, random_state=42)
ab.fit(X_train, y_train)

y_pred_ab = ab.predict(X_test)
y_proba_ab = ab.predict_proba(X_test)[:, 1]

print("AdaBoost Performance:")
print(f"Accuracy:  {accuracy_score(y_test, y_pred_ab):.4f}")
print(f"Precision: {precision_score(y_test, y_pred_ab):.4f}")
print(f"Recall:    {recall_score(y_test, y_pred_ab):.4f}")
print(f"F1 Score:  {f1_score(y_test, y_pred_ab):.4f}")
print(f"ROC-AUC:   {roc_auc_score(y_test, y_proba_ab):.4f}")

AdaBoost Performance:
Accuracy:  0.6189
Precision: 0.6000
Recall:    0.0703
F1 Score:  0.1259
ROC-AUC:   0.5823


OVERALL ANALYSIS

| Model                    | Accuracy | Precision | Recall | F1 Score | ROC-AUC |
| ------------------------ | -------- | --------- | ------ | -------- | ------- |
| Logistic Regression      | 0.6098   | 0.0000    | 0.0000 | 0.0000   | 0.5484  |
| Logistic Regression (L1) | 0.6098   | 0.0000    | 0.0000 | 0.0000   | 0.5485  |
| Logistic Regression (L2) | 0.6098   | 0.0000    | 0.0000 | 0.0000   | 0.5484  |
| KNN Classifier           | 0.6113   | 0.5031    | 0.3203 | 0.3914   | 0.5991  |
| SVM Classifier           | 0.6692   | 0.6970    | 0.2695 | 0.3887   | 0.6481  |
| Decision Tree            | 0.6326   | 0.6119    | 0.1602 | 0.2539   | 0.5656  |
| Random Forest            | 0.6555   | 0.6154    | 0.3125 | 0.4145   | 0.6430  |
| Naive Bayes              | 0.6143   | 0.5161    | 0.1875 | 0.2751   | 0.6088  |
| Gradient Boosting        | 0.6509   | 0.6364    | 0.2461 | 0.3549   | 0.6560  |
| AdaBoost                 | 0.6189   | 0.6000    | 0.0703 | 0.1259   | 0.5823  |

In [87]:

import pickle

filename = "water_potability_svm.sav"

with open(filename, "wb") as f:
    pickle.dump((model, scaler), f)

with open(filename, "rb") as f:
    loaded_model, loaded_scaler = pickle.load(f)

print("Model and Scaler loaded successfully!")

Model and Scaler loaded successfully!
